File: libs/power-log-parser/src/lib/parsers/battlegrounds-minions-bought-parser.ts
import { BlockType, CardIds, GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Action, Node, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BattlegroundsMinionsBoughtParser implements ActionParser {
	readonly ParserName = 'BattlegroundsMinionsBoughtParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	private Cards: string[] = [
		CardIds.DragToBuy,
		CardIds.DragToBuySpell_TB_BaconShop_DragBuy_Spell,
	];

	constructor(parserState: ParserState, facade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = facade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			this.StateFacade.IsBattlegrounds() &&
			node.Type === Action &&
			(node.Object as Action).Type === (BlockType.POWER as number) &&
			this.GameState.CurrentEntities.has((node.Object as Action).Entity) &&
			this.Cards.includes(this.GameState.CurrentEntities.get((node.Object as Action).Entity)!.CardId)
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		const controllerTagChange = action.Data.filter((data) => data instanceof TagChange)
			.map((data) => data as unknown as TagChange)
			.find((tag) => tag.Name === (GameTag.CONTROLLER as number));
		if (controllerTagChange == null) {
			return null;
		}

		const boughtMinion = this.GameState.CurrentEntities.get(controllerTagChange.Entity)!;
		const cardId = boughtMinion.CardId;
		const controller = controllerTagChange.Value;
		return [
			GameEventProvider.Create(
				(node.Object as Action).TimeStamp,
				'BATTLEGROUNDS_MINION_BOUGHT',
				GameEventHelper.CreateProvider(
					'BATTLEGROUNDS_MINION_BOUGHT',
					cardId,
					controller,
					boughtMinion.Entity,
					this.StateFacade,
					null,
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsMinionsBoughtParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using static HearthstoneReplays.Events.CardIds;
using System;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsMinionsBoughtParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        private List<string> Cards = new List<string>()
        {
            CardIds.DragToBuy,
            CardIds.DragToBuySpell_TB_BaconShop_DragBuy_Spell,
        };

        public BattlegroundsMinionsBoughtParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && StateFacade.IsBattlegrounds()
                && node.Type == typeof(Action)
                && (node.Object as Action).Type == (int)BlockType.POWER
                && GameState.CurrentEntities.ContainsKey((node.Object as Action).Entity)
                && Cards.Contains(GameState.CurrentEntities[((node.Object as Action).Entity)].CardId);
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var action = node.Object as Action;
            // There are some empty buy actions, not exactly sure what they are about
            // So we only count an action if a minion actually changed controllers
            var controllerTagChange = action.Data
                .Where(data => data is TagChange)
                .Select(data => data as TagChange)
                .Where(tag => tag.Name == (int)GameTag.CONTROLLER)
                .FirstOrDefault();
            if (controllerTagChange == null)
            {
                return null;
            }

            var boughtMinion = GameState.CurrentEntities[controllerTagChange.Entity];
            var cardId = boughtMinion.CardId;
            var controller = controllerTagChange.Value;
            return new List<GameEventProvider> { GameEventProvider.Create(
                    (node.Object as Action).TimeStamp,
                     "BATTLEGROUNDS_MINION_BOUGHT",
                    GameEvent.CreateProvider(
                        "BATTLEGROUNDS_MINION_BOUGHT",
                        cardId,
                        controller,
                        boughtMinion.Entity,
                        StateFacade,
                        null),
                    true,
                    node)
                };
        }
    }
}
